### ISSUE TITLE
Bug: `on: { '*': ... }` typing has broken after upgrade to v4.32

### ISSUE THREAD

**Issue by huan** [statelyai/xstate#3274]
### Description

My mailbox project has broken after upgrading XState from v4.30 to v4.32

### Expected result

Works as before

### Actual result

```
(property) StateNodeConfig<Context, any, EmptyAction<"mailbox/ACTOR_IDLE"> | PayloadAction<"mailbox/ACTOR_REPLY", { message: EventObject; }> | PayloadAction<...> | PayloadAction<...> | PayloadAction<...> | { ...; }, BaseActionObject>.on?: TransitionsConfig<context.Context, EmptyAction<"mailbox/ACTOR_IDLE"> | PayloadAction<"mailbox/ACTOR_REPLY", {
    message: EventObject;
}> | PayloadAction<...> | PayloadAction<...> | PayloadAction<...> | {
    ...;
}> | undefined
The mapping of event types to their potential transition(s).

Type '{ "mailbox/ACTOR_IDLE": { actions: any[]; target: string; }; "mailbox/ACTOR_REPLY": { actions: any[]; }; }' is not assignable to type 'TransitionsConfig<Context, EmptyAction<"mailbox/ACTOR_IDLE"> | PayloadAction<"mailbox/ACTOR_REPLY", { message: EventObject; }> | PayloadAction<...> | PayloadAction<...> | PayloadAction<...> | { ...; }> | undefined'.
  Type '{ "mailbox/ACTOR_IDLE": { actions: any[]; target: string; }; "mailbox/ACTOR_REPLY": { actions: any[]; }; }' is not assignable to type 'undefined'.ts(2322)
types.d.ts(339, 5): The expected type comes from property 'on' which is declared here on type 'StateNodeConfig<Context, any, EmptyAction<"mailbox/ACTOR_IDLE"> | PayloadAction<"mailbox/ACTOR_REPLY", { message: EventObject; }> | PayloadAction<...> | PayloadAction<...> | PayloadAction<...> | { ...; }, BaseActionObject>'
```

### Reproduction

https://github.com/huan/mailbox/blob/9a1a72ee189c0c465d34363d5891b250f020fb90/src/wrap.ts#L110-L113

### Additional context

The reproduction link is the code that I'm testing today.

The repo will be able to reproduce the typing problem by running:

```sh
npm install
npm run lint
```

**Comment by davidkpiano** [comment]
cc. @Andarist 

Possibly related to https://github.com/statelyai/xstate/pull/3252

**Comment by huan** [comment]
Additional information:

## This does not work

```ts
    on: {
      '*': {
        actions: context.queue.newMessage(actorMachine.id)(normalizedOptions.capacity) as any,
      },
    },
```

## This works

```ts
    on: {
      '*': {
        actions: context.queue.newMessage(actorMachine.id)(normalizedOptions.capacity),
      } as any,
    },
```

So it's wired that I can not understand why "xxx  is not assignable to type 'undefined'."

**Comment by huan** [comment]
After the above-linked commit, I can confirm that:

1. the behavior changed between v4.31 to v4.32
2. with the v4.31, I can use the `Workaround 1` above to make it works; but it does not work anymore with v4.32

**Comment by Andarist** [comment]
This is a backlash from https://github.com/statelyai/xstate/pull/3228/files . So far I've diagnosed this to be related to using generic `TEvent["type"]` here. I've created a minimal repro demonstrating the problem [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgUQG4FMB2MDyARgFZYDGMANCnACpQCGOAzsDMBDgMKcBmwAc3xRaDKAKxUaaAB4xGFTLimoAgjgCeSvEVIVqqLgAsIEZllUUOOagF84vKBBBwARDOYwGMLK4DccAD0gXAwGmBYzABccAAsAHQAzABM8QAMSEhhEXSMLGzWzDw4-AIAsgxgADy0xT5ylHTa8FhyuAAmzOjYOiTkMAB8cAC8iDQA2gDScMA4TT0w467ZvgC6cAA+bq6bbgBUrqsA-DHTrT44ndu7rgc0qEe5TKzsnMWlwqLikjV15420ZoDe5wGL0Z4FN58QSfMQSGC-Tj1KjoOQKGDNGrNRqTAYDPxIWwErLhLBPfKvFjvQQVMD4AA27UReH+82UcHOHS6zV0-SGo2QqCmMzmgIWSxWh2OYLyL2s1KEIjhPxBtSR-1VzQ5bUuXQQoVJpzg9keYvZMRwWGwUBo+MJcAAZGNUK5XCdyXKoSUYUrvgi1SyGmy8Hb7E7Bfs3TKIZSFbC-czkQCgQSiZlWpBYA4AK44KycOASS1QYBkABiOCxC21FyuPL6FAGAAoAJTOuBkTieUKRGDRimFBW0qog-Wd3M+KAW7MgQhYKDGgyoXb6lYxVwASQAclwdrYQcFQkZgF1c9Y4CAsLP5zMumQGNnWDgBHAGHAwE5CPSr3AjPOsCCWyroawaLMspKHMatojIg9gMHqIK3K4MT6gw+YsDEximOYliUlUY4QBO87TteC62I06haAsvKNsaBKoGmqBQJI2ZQHMPieKmSBAA). 

The interesting part is that even with the old version of thus type the assignability problem **doesn't go away**. It goes away in the more complex situation that involved `createMachine` call though. This leads me to believe that perhaps this was never OK - but TS just allowed it accidentally/silently in this more complex scenario. I will take another look at this in the upcoming week.
